control_protocol.md

```diff
@@ -0,0 +1,359 @@
+The {ref]}`message-layer` is the common language to understand commands, thus creating a remote procedure call.
+
+
+(transport-layer)=
+## Transport layer
+
+The transport layer ensures, that a message arrives its destination.
+
+
+### Sockets and Connections
+
+We use [zmq](https://zeromq.org/) sockets for our communication. For more details see the [zmq guide](https://zguide.zeromq.org/) or [zmq API](http://api.zeromq.org/)
+
+
+#### Definitions
+
+Zmq messages consist in a series of frames, each is a byte sequence.
+We indicate the separation between frames with `|`.
+An empty frame is indicated with two frame separators `||`, even at the beginning or end of a message.
+For example `||Second frame|Third frame||Fifth frame` consists of 5 frames.
+The first and fourth frames are empty frames.
+
+
+#### Configuration
+
+Each {ref}`Coordinator <components.md#coordinator>` shall offer one ROUTER socket, bound to a host name (or IP address or any address of a computer with "*") and port.
+
+{ref}`Components <components.md#components>` shall have one DEALER socket connecting to one Coordinator's ROUTER socket.
+
+Coordinators shall have one DEALER socket per other Coordinator in the Network.
+This DEALER socket shall connect to the other Coordinator's ROUTER socket.
+
+Messages must be sent to a Coordinator's ROUTER socket.
+
+
+(router-sockets)=
+#### Particularities of ROUTER sockets
+
+A small introduction to ROUTER sockets, for more details see [zmq guide chapter 3](https://zguide.zeromq.org/docs/chapter3/#Exploring-ROUTER-Sockets).
+
+A router socket assigns a random _identity_ to each connecting peer.
+If, for example, two Components `CA`, `CB` connect to the ROUTER sockets, the socket assigns the identities, which we call `IA`, `IB` (they can be any bytes sequence).
+
+Whenever a message is sent to the ROUTER socket from a peer, it reads as a first frame that identity.
+For example if `CA` sends the message `Request A`, the ROUTER socket will read `IA|Request A`.
+That way, you can return an answer to exactly the original peer and not, for example `CB`.
+If you call the ROUTER's send command with `IA|Reply A`, the socket will send `Reply A` to the peer, whose connection is `IA`, in this case `CA`.
+
+The following diagram shows this example communication with two Components:
+sequenceDiagram
+    participant Code as Message handling
+    participant ROUTER as ROUTER socket
+    Note over Code, ROUTER: Coordinator
+    CA ->> ROUTER: "Request A"
+    ROUTER ->> Code: "IA|Request A"
+    Code ->> ROUTER: "IA|Reply A"
+    ROUTER ->> CA: "Reply A"
+    CB ->> ROUTER: "Request B"
+    ROUTER ->> Code: "IB|Request B"
+    Code ->> ROUTER: "IB|Reply B"
+    ROUTER ->> CB: "Reply B"
+:::
+
+
+### Protocol basics
+
+#### Naming scheme
+
+Each Component has an individual name, given by the user, the _Component Name_.
+A Component Name must be a series of bytes, without the ASCII character "." (byte value 46).
```

Comment:
It is easier for the coordinator, if it does not have to decode at all. Therefore the Coordinator splits the byte sequence at a byte with value 46 and uses the first part as namespace and the second part as Component name.

You can use ASCII encoding, but you do not have to.

> Whitespace allowed?

So, yes, whitespaces and values above 127 are allowed.

> Although here we should probably not talk about bytes at all, and just define what the permissible character range is (ASCII or UTF-8, except ".")

So, you're for using some encoding?

Reply:
Well, I imagined (for readability/debugging alone) to store the names as strings (`Lab.MyDevice` vs. `4c 61 62 2e 4d 79 44 65 76 69 63 65`), but of course you're right, it will be faster to compare the raw byte values to received frames, and for the computer it doesn't matter. Over the wire and I guess in the zmq frames we'll have bytes only.
However, all user-facing stuff (e.g. logs, printouts, debugging, interfaces) should use strings.

If we stick to printable ASCII as the allowed character set, the en/decoding will be quite simple. UTF-8, less so I imagine.

> So, yes, whitespaces and values above 127 are allowed.

Now I'm confused. What's the character for a code point of 128 (byte 80)? Do you mean UTF-8?
What if we stick to printable ASCII (0x20 to 0x7E)?

Reply:
> Now I'm confused. What's the character for a code point of 128 (byte 80)? Do you mean UTF-8?
What if we stick to the printable ASCII range (0x20 to 0x7E, excluding 0x2E) for the names?

I said any bytes value, therefore also whitespaces. Whether they are useful for a user is another question.

So, let us change it to: 'Names consist of ASCII characters excluding the period (".")'.
If a user want stupid names (including space, new line, carriage return...), its his problem.

> strings

Strings is not that easy. For example a string in C has to be zero terminated. If we stick to "byte sequence representing ASCII characters", everybody knows what to expect and what not.

Reply:
> If a user want stupid names (including space, new line, carriage return...), its his problem.

It becomes our problem as soon as someone starts shenanigans with "funny" names.

> A Component name must be a series of ASCII characters, without the character "."

In principle fine, I'd like to restrict this to printable ASCII, but this is an encapsulated/independent change -- I'll open an issue and resolve here.